pypdf/_text_extraction/_layout_mode/_fixed_width_page.py

```diff
@@ -0,0 +1,398 @@
+                if _text:
+                    bt_groups.append(bt_group(tj_ops[0], _text, displaced_tx))
+                xform_stack.reset_tm()
+            return bt_groups, tj_ops
+        if op == b"q":
+            bts, tjs = recurs_to_target_op(ops, xform_stack, b"Q", fonts, debug)
+            bt_groups.extend(bts)
+            tj_ops.extend(tjs)
+        elif op == b"cm":
+            xform_stack.add_cm(*opands)
+        elif op == b"BT":
+            bts, tjs = recurs_to_target_op(ops, xform_stack, b"ET", fonts, debug)
+            bt_groups.extend(bts)
+            tj_ops.extend(tjs)
+        elif op == b"Tj":
+            tj_ops.append(decode_tj(opands[0], xform_stack))
+        elif op == b"TJ":
+            _tj = xform_stack.text_state_params()
+            for tj_op in opands[0]:
+                if isinstance(tj_op, bytes):
+                    _tj = decode_tj(tj_op, xform_stack)
+                    tj_ops.append(_tj)
+                else:
+                    xform_stack.add_trm(_tj.displacement_matrix(TD_offset=tj_op))
+        elif op == b"'":
+            xform_stack.reset_trm()
+            xform_stack.add_tm([0, xform_stack.TL])
+            tj_ops.append(decode_tj(opands[0], xform_stack))
+        elif op == b'"':
+            xform_stack.reset_trm()
+            _set_state_param(b"Tw", [opands[0]], xform_stack)
+            _set_state_param(b"Tc", [opands[1]], xform_stack)
+            xform_stack.add_tm([0, xform_stack.TL])
+            tj_ops.append(decode_tj(opands[2], xform_stack))
+        elif op in (b"Td", b"Tm", b"TD", b"T*"):
+            xform_stack.reset_trm()
+            if op == b"Tm":
+                xform_stack.reset_tm()
+            elif op == b"TD":
+                _set_state_param(b"TL", [-opands[1]], xform_stack)
+            elif op == b"T*":
+                opands = [0, -xform_stack.TL]
+            xform_stack.add_tm(opands)
+        elif op == b"Tf":
+            xform_stack.font_size = opands[1]
+            xform_stack.font = fonts[opands[0]]
+        else:
+            _set_state_param(op, opands, xform_stack)
+
+
+def y_coordinate_groups(
+    bt_groups: List[BTGroup], debug_path: Union[Path, None] = None
+) -> Dict[int, List[BTGroup]]:
+    """
+    Group text operations by rendered y coordinate, i.e. the line number
+
+    Args:
+        bt_groups: list of dicts as returned by text_show_operations()
+        debug_file: full path + filename prefix for debug output. Defaults to None.
+
+    Returns:
+        Dict[int, List[BTGroup]]: dict of lists of text rendered by each BT operator
+            keyed by y coordinate
+    """
+    ty_groups = {
+        ty: sorted(grp, key=lambda x: x["tx"])
+        for ty, grp in groupby(
+            bt_groups, key=lambda bt_grp: int(bt_grp["ty"] * bt_grp["flip_sort"])
+        )
+    }
+    # combine groups whose y coordinates differ by less than the effective font height
+    # (accounts for mixed fonts and other minor oddities)
```

Comment:
I've not tested a page rotated by 90<sup>o</sup> and honestly considered that out of scope (same for independently rotated text). The use case I'm targeting is "business-y" (think reporting engine PDFs), so at the end of the day, I'd direct users to the existing "plain" text extraction mode for exotic rotation use cases...  Thoughts?

Reply:
NOTE: combined landscape / portrait pages are fine.

Reply:
I've added rotation handling insofar as by default, if text is rotated w.r.t. the page, it will be excluded from the layout mode output. If rotated text is discovered, the user will receive a logger_warning "Rotated text discovered. Output will be incomplete."  The exclusion behavior can be overridden by setting extract_text() kwarg "layout_mode_strip_rotated" to False in which case ALL text is returned and the logger_warning states "Rotated text discovered. Layout will be degraded."